arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EffectScope.kt

```diff
@@ -240,3 +240,17 @@ public suspend fun <R, B : Any> EffectScope<R>.ensureNotNull(value: B?, shift: (
   contract { returns() implies (value != null) }
   return value ?: shift(shift())
 }
+
+/**
+ * Catch any possible `shift`ed outcome from [action], and transform it with [handler].
+ * This reads like a `catch` block, but for `EffectScope`.
+ *
+ * The [handler] may `shift` into a different `ErrorScope`, which is useful to
+ * simulate re-throwing of exceptions.
+ */
+public suspend fun<E, R, A> EffectScope<E>.catch(
```

Comment:
Needs to be an extension due to variance?

This duplicates behaviour with [`handleErrorXXX`](https://github.com/arrow-kt/arrow/blob/051847de3b1d3186fc9843864aaf642a2f4a854f/arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt#L661) but it exposes a nicer lambda for `handler` where you can leverage the DSL again by exposing `EffectScope<E>` on the receiver of the lambda.

I am absolutely up for improving these APIs, and renaming them but we should probably avoid duplicated APIs as much as possible to keep the API surface reasonable.

That being said this signatures reminds me a lot of [`Flow#catch`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/catch.html), and it uses the same naming so I am in favour of `catch` over the current `handleErrorXXX` & `redeemXXX` methods. This seems to cover all 4 APIs with a single nice API.

Reply:
About not being an extension, I just wrote it directly like that, although it could also be part of `EffectScope` itself. In fact, this opens the decades-old debate: it `catch` part of the "error effect", or is it a "effect transformer" (my definition of `catch` says that the latter).

Reply:
This also an alternative, using the `UnsafeVariance`, which is fine here.

// in EffectScope interface
```kotlin
public suspend fun <B> Effect<R, B>.catch(recover: suspend EffectScope<R>.(@UnsafeVariance R) -> B): B =
    fold({ recover(it) }, ::identity)
```